### PR TITLE
[cxx] Make some code valid C++, esp. adding casts and esp. not touching allocators. alternative to 10036

### PR DESCRIPTION
**Pull request by jaykrell**
This is an alternative to https://github.com/mono/mono/pull/10036.
Instead of macros, or casts, it uses templates.
The line damage of the casts does not seem good to me and I prefer templates or macros.

### CODE REVIEW

File: mono/metadata/threads-types.h [hunk 1]
```diff
@@ -117,7 +119,7 @@ ICALL_EXPORT
 int ves_icall_System_Threading_Thread_GetPriority (MonoThreadObjectHandle this_obj, MonoError *error);
 
 ICALL_EXPORT
-void ves_icall_System_Threading_Thread_SetPriority (MonoThreadObjectHandle this_obj, int priority, MonoError *error);
+void ves_icall_System_Threading_Thread_SetPriority (MonoThreadObjectHandle this_obj, MonoThreadPriority priority, MonoError *error);
```
Comment (luhenry) [inline mono/metadata/threads-types.h]: You have no guarantee on the size of the enum in Native, while managed will always pass an int. Revert that change.

**Comment by jaykrell** [comment]
`enum` will be `int` or `unsigned` here but ok.

**Comment by jaykrell** [comment]
Another tradeoff is `typedef MonoThreadPriority int`.
It loses typesafety and debugging, but works, is compatible and kinda looks nice like the previous.
C already had no typesafety here. But I'll just put it back.

**Comment by jaykrell** [comment]
Also `loader.h` don't touch, move to `loader-internals.h` or other.

**Comment by jaykrell** [comment]
@monojenkins build Linux AArch64 Coop Suspend

**Comment by jaykrell** [comment]
@monojenkins build Linux AArch64 FullAOT

**Comment by jaykrell** [comment]
@monojenkins build OS X i386

File: mono/mini/aot-runtime.c [hunk 2]
```diff
@@ -1583,7 +1583,7 @@ aot_cache_load_module (MonoAssembly *assembly, char **aot_name)
 		if (!home)
 			return NULL;
 		cache_dir = g_strdup_printf ("%s/Library/Caches/mono/aot-cache", home);
-		if (!g_file_test (cache_dir, G_FILE_TEST_EXISTS|G_FILE_TEST_IS_DIR))
+		if (!g_file_test (cache_dir, (GFileTest)(G_FILE_TEST_EXISTS|G_FILE_TEST_IS_DIR)))
```
Comment (jaykrell) [inline mono/mini/aot-runtime.c]: This cast is redundant with #10274 or #10231.

**Comment by jaykrell** [comment]
@monojenkins build OS X i386

**Comment by jaykrell** [comment]
@monojenkins build Linux AArch64 Coop Suspend

**Comment by jaykrell** [comment]
@monojenkins build OS X i386

**Comment by jaykrell** [comment]
@monojenkins build failed

**Comment by jaykrell** [comment]
Good now?

**Comment by jaykrell** [comment]
@luhenry ok now?

**Comment by jaykrell** [comment]
@luhenry ok now?

**Comment by jaykrell** [comment]
Review?

File: mono/metadata/class-internals.h [hunk 3]
```diff
@@ -1067,6 +1067,22 @@ mono_register_jit_icall (gconstpointer func, const char *name, MonoMethodSignatu
 MonoJitICallInfo *
 mono_register_jit_icall_full (gconstpointer func, const char *name, MonoMethodSignature *sig, gboolean no_wrapper, const char *c_symbol);
 
+#ifdef __cplusplus
+template <typename T>
+inline MonoJitICallInfo *
+mono_register_jit_icall (T func, const char *name, MonoMethodSignature *sig, gboolean is_save)
```
Comment (lambdageek) [inline mono/metadata/class-internals.h]: One day, maybe, we can make this more typesafe than the C version:

```c++
template <typename T>
inline typename std::enable_if<std::is_function <T>::value, MonoJitICallInfo*>::type
mono_register_jit_icall (T func, const char *name, MonoMethodSignature *sig, gboolean is_save)
{
   return mono_register_jit_icall ((gconstpointer)func, name, sig, is_save);
}
```

File: mono/mini/aot-runtime.c [hunk 4]
```diff
@@ -3114,7 +3114,7 @@ decode_exception_debug_info (MonoAotModule *amodule, MonoDomain *domain,
 			}
 		}
 
-		flags |= JIT_INFO_HAS_UNWIND_INFO;
+		flags = (MonoJitInfoFlags)(flags | JIT_INFO_HAS_UNWIND_INFO);
```
Comment (lambdageek) [inline mono/mini/aot-runtime.c]: still needed after `G_ENUM_FUNCTIONS (MonoJitInfoFlags)` ?

File: mono/metadata/class-internals.h (hunk 3 above)
Comment (jaykrell) [reply mono/metadata/class-internals.h]: Agreed but not today.
Does it end up being templates all the way down, or is this the correct C++ approach?

File: mono/mini/aot-runtime.c (hunk 4 above)
Comment (jaykrell) [reply mono/mini/aot-runtime.c]: Good catch. Unlikely still needed.

**Comment by jaykrell** [comment]
Undo the two lines right?

File: mono/metadata/class-internals.h (hunk 3 above)
Comment (lambdageek) [reply mono/metadata/class-internals.h]: in the libraries that i've seen, sooner or later the templates cease and you end up at some untyped storage layer where everything's a `char*` (and a destructor).

So for something that's essentially a big lookup table of heterogeneous functions this is probably close to the right place to cut down to a `gconstpointer`.

**Comment by lambdageek** [comment]
No need to cleanup the two casts now (look at that all-green :heavy_check_mark: CI!) - the casts aren't _wrong_, after all.

File: mono/mini/aot-runtime.c (hunk 4 above)
Comment (jaykrell) [reply mono/mini/aot-runtime.c]: https://github.com/mono/mono/issues/10319 (can undo this line later)

File: mono/mini/aot-runtime.c (hunk 2 above)
Comment (jaykrell) [reply mono/mini/aot-runtime.c]: https://github.com/mono/mono/issues/10319 (can undo this line later)

**Comment by jaykrell** [comment]
@lambdageek says elsewhere: "Oops... #10252 loader-internals.h needs to go into some csproj so it gets picked up on windows."

Yes, but why did it pass CI then?

**Comment by jaykrell** [comment]
Working on loader-internals.h right away though. I know where you mean.

**Comment by jaykrell** [comment]
Tarball build will be broken, again.

**Comment by jaykrell** [comment]
er, tarball maybe ok, it is in Makefile.am.in.  phew

**Comment by jaykrell** [comment]
Hm. Maybe not a problem?
```
jaymac1:msvc jay$ grep loader.h *
libmonoruntime-common.targets:    <ClInclude Include="$(MonoSourceLocation)\mono\metadata\loader.h" />
libmonoruntime-common.targets.filters:    <ClInclude Include="$(MonoSourceLocation)\mono\metadata\loader.h">
$ grep  internal  *proj*
```

**Comment by jaykrell** [comment]
Grep for `ClInclude`, some found, but not much.  So I think ok.

**Comment by jaykrell** [comment]
Oh, not proj, but common.targets.

**Pull request merged**